### PR TITLE
1/12 Bpm rounding

### PR DESCRIPTION
**Pull request by daschuer**
This uses the BeatUtils::roundBpmWithinRange() function for bpm adjusting and tabbing. 

Bpm adjustment is now done in 1/48 steps, reaching full integer BPM; 
Tabing is rounded to 1/12, which allows to tab full integer BPM with a bit practice.

### CODE REVIEW

**Comment by daschuer** [comment]
@poelzi: Did this work for you or is there anything to improve?

**Comment by Holzhaus** [comment]
ping @poelzi

**Comment by Holzhaus** [comment]
@daschuer Do you expect any fallout from this? What was the previous tapping resolution?
I can test this if you want.

**Comment by daschuer** [comment]
Before, the was no rounding during tapping.

**Comment by daschuer** [comment]
The other thing that can be tested is shrinking and widening the beat grid using the controls next to the waveform. 
Is it too coarse / fine?

**Comment by ronso0** [comment]
With the previous tapping algorhythm you could _never_ tap rounded integers -- even if it matched once you would mess it up again with just one more tap
https://bugs.launchpad.net/mixxx/+bug/1882776

**Comment by Holzhaus** [comment]
Is it just me or do the bpm adjustment buttons don't work anymore? Shifting the beatgrid works, but shrinking/growing doesn't make any difference even after pressing the button 20+ times (not even when I scroll to the track end, where changes should be more obvious).

Re: Tapping. Do we have a GUI button for that? If so, I can't find it.

**Comment by JoergAtGithub** [comment]
> 
> Re: Tapping. Do we have a GUI button for that? If so, I can't find it.

Click on the BPM value

**Comment by Holzhaus** [comment]
Ok, in that case BPM tapping is broken too and clicking the bpm value repeatedly has no effect.

**Comment by ronso0** [comment]
Tapping works fine for me (LateNight) and I can edit the beatgrid as before.

But indeed `beats_adjust_faster/slower` is to coarse now when you need to correct a track that's just slightly off integer BPM

**Comment by daschuer** [comment]
I have restored the 0.01 step width. 

But playing with it, I think the feature does not really match the use case. It has never the right step width and the effect depends strong to the bpm and the length of the track. 

So I think we can do better. (Not in this PR). 

My use case for const tracks is the following:
* Verify a the first beat
* adjust the offset
* Seek through the track towards the end 
* Verify if the markers are still on the beats. 
* Adjust the BPM 

Issues:
* In case it is fully off and the track is short the steps are too fine 
* In case the it is only slightly off and the rack is long, the steps are too big.

What I really want: 
* Set the a beat marker and adjust the BPM to match my first and my last marker.  
* I think we can ditch the BPM adjustment buttons for this. 

What do you think?

**Comment by Holzhaus** [comment]
I have no strong opinion on this as the BPM Analyzer mostly does its job correctly for my tracks and I usually only see cases where boom/tshack is misdetected (so I usually match phase manually and then apply the phase correction from the other deck).

But I agree that this "keyframe-like" workflow would be much more straighforwaed. IMHO we shouldn't invest too much time in this now. We have the #2961, which we should focus on, and we already have a lot of merge conflicts. Let's get that PR into a mergeable state and then follow up with the improved workflow (setting beat markers and interpolation beats in between) for adjusting beatgrids.

**Comment by daschuer** [comment]
Yes, right at least this it not a 2.3 topic, like this PR.

**Comment by ronso0** [comment]
@Holzhaus so you can tap and adjust with this branch?
I'll test again tonight if necessary.

**Comment by ronso0** [comment]
so this also fixes https://bugs.launchpad.net/mixxx/+bug/1882776

**Comment by ronso0** [comment]
ux still LGTM

**Comment by Holzhaus** [comment]
> Is it just me or do the bpm adjustment buttons don't work anymore? Shifting the beatgrid works, but shrinking/growing doesn't make any difference even after pressing the button 20+ times (not even when I scroll to the track end, where changes should be more obvious).

> BPM tapping is broken too and clicking the bpm value repeatedly has no effect.

I was wrong. Both work fine **if (and only if)** the track has a beatgrid with *constant tempo*. I accidently tested on a track where "Assume constant tempo" was not checked, so it didn't work and nothing happens.

Is it is feasible to disable (grey out) the beatgrid adjustment buttons? It's pretty non-obvious that this doens't work because the track has a non-const beatgrid.

File: src/track/beatgrid.cpp [hunk 1]
```diff
@@ -3,16 +3,22 @@
...
-static const int kFrameSize = 2;
+namespace {
 
 struct BeatGridData {
     double bpm;
     double firstBeat;
 };
 
+constexpr int kFrameSize = 2;
+constexpr double kBpmScaleRounding = 0.001;
```
Comment (Holzhaus) [inline src/track/beatgrid.cpp:18]: Previously this was 1/48 in this PR, which is a multiple of 1/12. Is there a reason for this? If so, have you considered 1/960 ?

**Review by Holzhaus — APPROVED** [review]
LGTM, thank you.

File: src/track/beatgrid.cpp (hunk 1 above)
Comment (Holzhaus) [reply src/track/beatgrid.cpp:18]: Anyway, this can be done into a follow-up PR if needed. I'm going to merge this now.

**Pull request merged**